compiler/rustc_session/src/filesearch.rs

```diff
@@ -62,30 +61,127 @@ pub fn make_target_lib_path(sysroot: &Path, target_triple: &str) -> PathBuf {
     PathBuf::from_iter([sysroot, Path::new(&rustlib_path), Path::new("lib")])
 }
 
-/// This function checks if sysroot is found using env::args().next(), and if it
-/// is not found, uses env::current_exe() to imply sysroot.
-pub fn get_or_default_sysroot() -> PathBuf {
-    // Follow symlinks.  If the resolved path is relative, make it absolute.
-    fn canonicalize(path: PathBuf) -> PathBuf {
-        let path = fs::canonicalize(&path).unwrap_or(path);
-        // See comments on this target function, but the gist is that
-        // gcc chokes on verbatim paths which fs::canonicalize generates
-        // so we try to avoid those kinds of paths.
-        fix_windows_verbatim_for_gcc(&path)
+#[cfg(unix)]
+fn current_dll_path() -> Option<PathBuf> {
+    use std::ffi::{CStr, OsStr};
+    use std::os::unix::prelude::*;
+
+    unsafe {
+        let addr = current_dll_path as usize as *mut _;
+        let mut info = std::mem::zeroed();
+        if libc::dladdr(addr, &mut info) == 0 {
+            info!("dladdr failed");
+            return None;
+        }
+        if info.dli_fname.is_null() {
+            info!("dladdr returned null pointer");
+            return None;
+        }
+        let bytes = CStr::from_ptr(info.dli_fname).to_bytes();
+        let os = OsStr::from_bytes(bytes);
+        Some(PathBuf::from(os))
     }
+}
 
-    // Use env::current_exe() to get the path of the executable following
-    // symlinks/canonicalizing components.
-    fn from_current_exe() -> PathBuf {
-        match env::current_exe() {
-            Ok(exe) => {
-                let mut p = canonicalize(exe);
-                p.pop();
-                p.pop();
-                p
+#[cfg(windows)]
+fn current_dll_path() -> Option<PathBuf> {
```

Comment:
Can you return `Result<PathBuf, String>` instead so that `expect()` shows a useful error?

Reply:
> @ozkanonur can you run `x test src/tools/clippy` in the meantime and see if that passes? It will take a long time to compile unfortunately.

![rustc-clippy](https://user-images.githubusercontent.com/39852038/198388712-4ccd100e-e2b2-43f9-a4c0-2558f7591155.png)

All passed.